20240919-scheduler-failure-policy.md

```diff
@@ -0,0 +1,135 @@
+4. The workflow Actor Reminders will be updated so that they will be marked as only to be triggered once, with an appropriate retry failure policy when using Scheduler.
+
+
+## Design
+
+To begin with, we will support 3 failure policies:
+1. `Drop`: the job trigger will not be retried and the job will be marked as ticked.
+  This is the current behaviour and will continue to be the default behaviour for all jobs.
+2. `Constant`: the job will be retried as a constant time interval, up to a maximum number of retries (which could be infinite).
+3. `Schedule`: the job will be retried according to a [cron scheudler](https://github.com/diagridio/go-etcd-cron/blob/2a1c6747974627691165eb96a2ca0202285d71eb/proto/job.proto#L68), up to a maximum number of retries (which could be infinite).
+
+### Future Design
+
+Although not part of the proposal, in future we can extend Scheduler to include a staging queue which is dedicated for Jobs where no current stream implements its target.
+This addition would mean that the Jobs with no target are not needlessly being attempted to be triggered, freeing up main queue resources and preserving the intended failure policy.
+Upon trigger, if no target stream is connected, the job will be moved to this staging queue.
+In the event a stream implementing the target is connected, the Job can be moved to the main queue and triggered immediately or on its proper schedule.
+
+### go-etcd-cron
+
+To support the new `FailurePolicy` options, we first need to keep track of the current number of attempts the current Job count has been triggered on a particular tick.
+We also need to keep track of the time at which that attempt was made in order to calculate the next attempt time according to the `FailurePolicy` schedule.
+Like the trigger time, the last attempt time is the virtual _correct_ time that the attempt was to be made, rather than the observed wall clock time.
+This is to ensure durability of the Scheduler during events of slow down or down time/restarts.
+The attempts and last attempt time will added to the Counter proto message, and managed by the trigger queue manager.
+Attempts will be 0 and last attempted time will be `nil` for all new tick counts of a Job.
+
+```proto
+// Counter holds counter information for a given job.
+message Counter {
+  ...
+
+  // attempts is the number of times the job has been attempted to be
+  // triggered at this count.
+  uint32 attempts = 4;
+
+  // last_attempted is the timestamp the job was last attempted to be triggered.
+  optional google.protobuf.Timestamp last_attempted = 5;
+}
+```
+
+Below are the proto definitions for the new `FailurePolicy` options.
+Both constant and cron policies include an optional max retries option to limit the number of retries according to the number of attempts.
+If max retries is unset, the Job will be retried indefinitely.
+The failure policy message is added as an optional field to the Job message.
+If unset, the failure policy of a Job is `Drop`.
```

Comment:
What would be the implications of setting the default failure policy to a 1s constant retry with 3 max attempts? Sure I think it's a good prod to set your own retry policies but I'd love some useful defaults.

Reply:
Yes I think this might be interesting, and am inclined to agree this should be the default. I do not see this as necessarily being different behaviour as today, since the "fire and forget" strategy would be more covered by whether the job is set to the staging queue rather than failure policy on the trigger. @cicoyle I think this also covers your point of having a global sane default.

WDYT @artursouza @cicoyle

Reply:
I support this as being the default

Reply:
I have updated the default behavior to Constant with `1s` delay & 3 maximum retries.

Reply:
Swell, looks good to me 👌